File: Controllers/RssFeedStatsController.vb
Imports System
Imports System.Web
Imports System.Web.Http
Imports System.Collections.Generic
Imports System.Linq
Imports AttributeRouting.Web.Http
Imports System.Data.SqlClient
Imports System.Web.Script.Serialization
Imports log4net
Imports System.Threading
Imports System.Drawing
Imports System.IO
Imports Newtonsoft.Json
Imports System.Net
Imports System.Text

Namespace CalcmenuAPI
    Public Class RssFeedStatsController
        Inherits ApiController
        Private Shared ReadOnly Log As ILog = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType)


        <HttpGet> <[GET]("/api/rssfeedstats")> _
        Public Function GetRecipe() As Models.RSSFeedStats

            Dim RSSFeedStats As New Models.RSSFeedStats
            Try
                Using cmd As New SqlCommand()
                    With cmd
                        Using cn As New SqlConnection(ConnectionString)
                            Try
                                .Connection = cn
                                .CommandType = CommandType.StoredProcedure
                                .CommandText = "[dbo].[sp_GetRSSFeedsStats]"
                                cn.Open()

                                Using dr As SqlDataReader = cmd.ExecuteReader()
                                    '' Main
                                    If dr.HasRows Then
                                        While dr.Read
                                            RSSFeedStats.statsOverview = GetStr(dr("Overview"))
                                        End While
                                    End If

                                    '' Per Year
                                    dr.NextResult()
                                    If dr.HasRows Then
                                        Dim PerYear As New List(Of Models.StatsPerYear)
                                        While dr.Read
                                            Dim Years As New Models.StatsPerYear With {
                                                .DetailsPerYear = GetStr(dr("DetailsPerYear"))
                                                }
                                            PerYear.Add(Years)
                                        End While
                                        RSSFeedStats.statsPerYear = PerYear
                                    End If


                                    '' Per Month
                                    dr.NextResult()
                                    If dr.HasRows Then
                                        Dim PerMonth As New List(Of Models.StatsPerMonth)
                                        While dr.Read
                                            Dim Months As New Models.StatsPerMonth With {
                                                .DetailsPerMonth = GetStr(dr("DetailsPerMonth"))
                                                }
                                            PerMonth.Add(Months)
                                        End While
                                        RSSFeedStats.statsPerMonth = PerMonth
                                    End If


                                    '' Per Day
                                    dr.NextResult()
                                    If dr.HasRows Then
                                        Dim PerDay As New List(Of Models.StatsPerDay)
                                        While dr.Read
                                            Dim Days As New Models.StatsPerDay With {
                                                .DetailsPerDay = GetStr(dr("DetailsPerDay"))
                                                }
                                            PerDay.Add(Days)
                                        End While
                                        RSSFeedStats.statsPerDay = PerDay
                                    End If

                                    '' Per Hours
                                    dr.NextResult()
                                    If dr.HasRows Then
                                        Dim PerHour As New List(Of Models.StatsPerHours)
                                        While dr.Read
                                            Dim Days As New Models.StatsPerHours With {
                                                .DetailsPerHours = GetStr(dr("DetailsPerHour"))
                                                }
                                            PerHour.Add(Days)
                                        End While
                                        RSSFeedStats.statsPerHour = PerHour
                                    End If


                                    dr.Close()
                                End Using
                            Finally
                                If Not cn Is Nothing Then
                                    cn.Close()
                                    CType(cn, IDisposable).Dispose()
                                End If
                            End Try
                        End Using
                    End With
                End Using
            Catch aex As ArgumentException
                Log.Warn(System.Reflection.MethodBase.GetCurrentMethod().Name + ": Missing or invalid parameters", aex)
                Throw New HttpResponseException(GenericErrorResponse("Request failed", HttpStatusCode.BadRequest, 440))
            Catch hex As HttpResponseException
                Log.Error(System.Reflection.MethodBase.GetCurrentMethod().Name + ": Unexpected error occured", hex)
                Throw hex
            Catch ex As Exception
                Log.Error(System.Reflection.MethodBase.GetCurrentMethod().Name + ": Unexpected error occured", ex)
                Throw New HttpResponseException(GenericErrorResponse("Request failed", HttpStatusCode.InternalServerError, 500))
            End Try

            Return RSSFeedStats
        End Function


    End Class
End Namespace

File: src/CmcStandardPrinting.Api/Controllers/RssFeedStatsController.cs
using System;
using System.Collections.Generic;
using System.Data;
using Microsoft.Data.SqlClient;
using CmcStandardPrinting.Domain.RssFeeds;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace CmcStandardPrinting.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class RssFeedStatsController : ControllerBase
{
    private readonly IConfiguration _configuration;
    private readonly ILogger<RssFeedStatsController> _logger;

    public RssFeedStatsController(IConfiguration configuration, ILogger<RssFeedStatsController> logger)
    {
        _configuration = configuration;
        _logger = logger;
    }

    private string ConnectionString => _configuration.GetConnectionString("Default") ?? string.Empty;

    [HttpGet]
    public ActionResult<RssFeedStats> GetRecipe()
    {
        var stats = new RssFeedStats();
        try
        {
            using var cmd = new SqlCommand();
            using var cn = new SqlConnection(ConnectionString);
            cmd.Connection = cn;
            cmd.CommandType = CommandType.StoredProcedure;
            cmd.CommandText = "[dbo].[sp_GetRSSFeedsStats]";
            cn.Open();

            using var dr = cmd.ExecuteReader();
            if (dr.HasRows)
            {
                while (dr.Read())
                {
                    stats.StatsOverview = GetStr(dr["Overview"]);
                }
            }

            dr.NextResult();
            if (dr.HasRows)
            {
                var perYear = new List<StatsPerYear>();
                while (dr.Read())
                {
                    perYear.Add(new StatsPerYear
                    {
                        DetailsPerYear = GetStr(dr["DetailsPerYear"])
                    });
                }
                stats.StatsPerYear = perYear;
            }

            dr.NextResult();
            if (dr.HasRows)
            {
                var perMonth = new List<StatsPerMonth>();
                while (dr.Read())
                {
                    perMonth.Add(new StatsPerMonth
                    {
                        DetailsPerMonth = GetStr(dr["DetailsPerMonth"])
                    });
                }
                stats.StatsPerMonth = perMonth;
            }

            dr.NextResult();
            if (dr.HasRows)
            {
                var perDay = new List<StatsPerDay>();
                while (dr.Read())
                {
                    perDay.Add(new StatsPerDay
                    {
                        DetailsPerDay = GetStr(dr["DetailsPerDay"])
                    });
                }
                stats.StatsPerDay = perDay;
            }

            dr.NextResult();
            if (dr.HasRows)
            {
                var perHour = new List<StatsPerHours>();
                while (dr.Read())
                {
                    perHour.Add(new StatsPerHours
                    {
                        DetailsPerHours = GetStr(dr["DetailsPerHour"])
                    });
                }
                stats.StatsPerHour = perHour;
            }
        }
        catch (ArgumentException ex)
        {
            _logger.LogWarning(ex, "Missing or invalid parameters for RSS feed stats");
            return BadRequest();
        }
        catch (SqlException ex)
        {
            _logger.LogError(ex, "Database error while fetching RSS feed stats");
            return StatusCode(500);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Unexpected error while fetching RSS feed stats");
            return StatusCode(500);
        }

        return Ok(stats);
    }

    private static string GetStr(object? value)
    {
        if (value == null || value == DBNull.Value) return string.Empty;
        return value.ToString() ?? string.Empty;
    }
}
